src/mem/cache/prefetch/ipcp.cc

```diff
@@ -0,0 +1,398 @@
+           if (candidate->second.lastTouch
+            < victim->second.lastTouch)
+               victim = candidate;
+           if (candidate->second.trained) {
+               victim = candidate;
+               ipcpStats.trainedVictim += 1;
+               break;
+           }
+        }
+
+        if (victim->second.trained) {
+            for (auto &ip : ipcpTable) {
+                if (ip.second.regionAddr == victim->first) {
+                    ip.second.streamValid = true;
+                    ip.second.dir = (victim->second.dir > 32) ? 1 : -1;
+                }
+            }
+        }
+
+        RST.erase(victim);
+    }
+
+   //CS, CPLX
+
+    int64_t curStride = 0;
+
+    //Lookup IPCP
+    if (ipcpTable.find(ipIndex) != ipcpTable.end()) {
+        if (ipcpTable[ipIndex].ipTag == ipTag) {
+            //Hit in IPCP table
+            ipcpTable[ipIndex].valid = 1;
+        }
+        else {
+            //Check valid for taking replacement decision
+            if (ipcpTable[ipIndex].valid == 0) {
+                ipcpTable[ipIndex].ipTag = ipTag;
+                ipcpTable[ipIndex].lastVpage = curPage;
+                ipcpTable[ipIndex].lastLineOffset = lineOffset;
+                ipcpTable[ipIndex].signature = 0;
+                ipcpTable[ipIndex].stride = 0;
+                ipcpTable[ipIndex].conf = 0;
+                ipcpTable[ipIndex].valid = 1;
+                ipcpTable[ipIndex].regionAddr = regionBase;
+                ipcpTable[ipIndex].dir = 1;
+                ipcpTable[ipIndex].streamValid = (RST[regionBase].tentative
+                 || RST[regionBase].trained);
+                ipcpStats.conflicts += 1;
+
+            } else {
+                ipcpTable[ipIndex].valid = 0;
+            }
+        return; //Stop here if conflict
+        }
+    } else {
+        //New entry in IPCP Table
+        ipcpTable[ipIndex].ipTag = ipTag;
+        ipcpTable[ipIndex].lastVpage = curPage;
+        ipcpTable[ipIndex].lastLineOffset = lineOffset;
+        ipcpTable[ipIndex].signature = 0;
+        ipcpTable[ipIndex].stride = 0;
+        ipcpTable[ipIndex].conf = 0;
+        //ipcpTable[ipIndex].conf = 4;
+        ipcpTable[ipIndex].valid = 1;
+        ipcpTable[ipIndex].regionAddr = regionBase;
+        ipcpTable[ipIndex].dir = 1;
+        ipcpTable[ipIndex].streamValid = (RST[regionBase].tentative
+         || RST[regionBase].trained);
+    }
+
+    curStride = lineOffset - ipcpTable[ipIndex].lastLineOffset;
+    // Page boundary learning
+    if (ipcpTable[ipIndex].lastVpage != curPage) { //switching it off?
+       if (curStride < 0) //Assuming the page moving in +ve direction
+           curStride += blkSize;
+       else
+           curStride -= blkSize;
+    }
+    if (curStride == 0) return;
+
+    if (ipcpTable[ipIndex].stride == curStride) {
+        ipcpTable[ipIndex].conf += 1;
+    }
+    else
+        // should we return from here?
+        ipcpTable[ipIndex].conf -= 1;
```

Comment:
Required to have curly braces for multi-line if statements. I have lost many hours of my life tracking down bugs because people wanted to save two characters. It's not worth it.